File: ios/HPTravail/HPTravail/SalaryV2/EmployerGeneralReductionRateContext2026V2.swift
import Foundation

/// Contexte légal du coefficient maximal RGDU 2026.
///
/// D.241-7 CSS rattache Tδ au taux de contribution logement réellement dû
/// (L.813-5 1° ou 2° CCH), jamais au seul effectif brut. Si la somme des taux
/// éligibles effectivement à la charge de l'employeur est inférieure au maximum
/// standard, Tδ est réduit afin de ne pas dépasser cette somme.
enum EmployerGeneralReductionRateContext2026V2 {
    private static let tMin = 0.0200
    private static let tDeltaL813_5_1 = 0.3781
    private static let tDeltaL813_5_2 = 0.3821

    enum HousingContributionRegime: Equatable {
        /// CCH L.813-5 1° : 0,1 % sur assiette plafonnée.
        case l813_5_1
        /// CCH L.813-5 2° : 0,5 % sur totalité de l'assiette.
        case l813_5_2
    }

    struct Record: Equatable {
        let id: String
        let housingContributionRegime: HousingContributionRegime
        /// Somme décimale des taux RGDU éligibles effectivement à la charge de l'employeur.
        let eligibleEmployerRateSum: Double
        let effectiveFrom: YearMonthV2
        let effectiveTo: YearMonthV2?
        let source: String
    }

    struct Snapshot: Equatable {
        let tDelta: Double?
        let maximumCoefficient: Double?
        let source: String?
        let reliable: Bool
        let warnings: [String]
    }

    static func resolve(records: [Record], period: YearMonthV2) -> Snapshot {
        guard period.year == 2026 else {
            return blocked("RGDU : contexte de coefficient 2026 indisponible pour \(period.year).")
        }

        let malformed = records.filter { record in
            record.id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                record.source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
                (record.effectiveTo.map { $0 < record.effectiveFrom } ?? false) ||
                !validEligibleRateSum(record.eligibleEmployerRateSum)
        }
        guard malformed.isEmpty else {
            return blocked("RGDU 2026 : règle de coefficient incomplète ou incohérente ; aucun Tδ n'est supposé.")
        }

        let active = records.filter { record in
            period >= record.effectiveFrom && (record.effectiveTo == nil || period <= record.effectiveTo!)
        }
        guard !active.isEmpty else {
            return blocked(
                "RGDU 2026 : régime de contribution logement et somme des taux éligibles à confirmer pour \(String(format: "%02d/%04d", period.month, period.year))."
            )
        }
        guard active.count == 1, let selected = active.first else {
            return blocked("RGDU 2026 : plusieurs règles de coefficient se chevauchent ; calcul automatique bloqué.")
        }

        let standardDelta: Double
        switch selected.housingContributionRegime {
        case .l813_5_1:
            standardDelta = tDeltaL813_5_1
        case .l813_5_2:
            standardDelta = tDeltaL813_5_2
        }
        let standardMaximum = tMin + standardDelta
        let effectiveMaximum = min(standardMaximum, selected.eligibleEmployerRateSum)
        let effectiveDelta = effectiveMaximum - tMin

        guard effectiveDelta.isFinite, effectiveDelta >= 0 else {
            return blocked("RGDU 2026 : somme des taux éligibles incompatible avec Tmin ; calcul automatique bloqué.")
        }

        return Snapshot(
            tDelta: effectiveDelta,
            maximumCoefficient: effectiveMaximum,
            source: selected.source.trimmingCharacters(in: .whitespacesAndNewlines),
            reliable: true,
            warnings: []
        )
    }

    static func isUsable(_ snapshot: Snapshot?) -> Bool {
        guard let snapshot,
              snapshot.reliable,
              let source = snapshot.source,
              !source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              let delta = snapshot.tDelta,
              let maximum = snapshot.maximumCoefficient,
              delta.isFinite,
              delta >= 0,
              maximum.isFinite,
              maximum >= tMin,
              maximum <= 1 else { return false }

        return abs((tMin + delta) - maximum) < 0.000_000_1
    }

    private static func validEligibleRateSum(_ value: Double) -> Bool {
        value.isFinite && value >= tMin && value <= 1
    }

    private static func blocked(_ warning: String) -> Snapshot {
        Snapshot(
            tDelta: nil,
            maximumCoefficient: nil,
            source: nil,
            reliable: false,
            warnings: [warning]
        )
    }
}

File: app/src/main/java/com/amaury/pointage/v2/engine/EmployerGeneralReductionRateContext2026V2.kt
package com.amaury.pointage.v2.engine

import java.time.YearMonth
import kotlin.math.min

/**
 * Contexte légal du coefficient maximal RGDU 2026.
 *
 * D.241-7 CSS rattache Tδ au taux de contribution logement réellement dû
 * (L.813-5 1° ou 2° CCH), pas au seul effectif brut de l'entreprise. Lorsque
 * la somme des taux effectivement à la charge de l'employeur est inférieure
 * au maximum standard, Tδ est réduit pour que le maximum RGDU ne dépasse pas
 * cette somme.
 *
 * Une règle doit donc être datée, sourcée et confirmer à la fois le régime
 * logement applicable et la somme des taux éligibles réellement supportés.
 */
object EmployerGeneralReductionRateContext2026V2 {
    private const val T_MIN = 0.0200
    private const val T_DELTA_L813_5_1 = 0.3781
    private const val T_DELTA_L813_5_2 = 0.3821

    enum class HousingContributionRegime {
        /** CCH L.813-5 1° : contribution logement au taux de 0,1 % sur assiette plafonnée. */
        L813_5_1,

        /** CCH L.813-5 2° : contribution logement au taux de 0,5 % sur la totalité de l'assiette. */
        L813_5_2
    }

    data class Record(
        val id: String,
        val housingContributionRegime: HousingContributionRegime,
        /**
         * Somme décimale des taux de cotisations/contributions éligibles RGDU
         * effectivement à la charge de l'employeur pour le salarié concerné.
         * Exemple : 0,4021 = 40,21 %.
         */
        val eligibleEmployerRateSum: Double,
        val effectiveFrom: YearMonth,
        val effectiveTo: YearMonth? = null,
        val source: String
    )

    data class Snapshot(
        val tDelta: Double?,
        val maximumCoefficient: Double?,
        val source: String?,
        val reliable: Boolean,
        val warnings: List<String>
    )

    fun resolve(records: List<Record>, period: YearMonth): Snapshot {
        if (period.year != 2026) {
            return blocked("RGDU : contexte de coefficient 2026 indisponible pour ${period.year}.")
        }

        val malformed = records.filter { record ->
            record.id.isBlank() ||
                record.source.isBlank() ||
                record.effectiveTo?.let { it < record.effectiveFrom } == true ||
                !validEligibleRateSum(record.eligibleEmployerRateSum)
        }
        if (malformed.isNotEmpty()) {
            return blocked("RGDU 2026 : règle de coefficient incomplète ou incohérente ; aucun Tδ n'est supposé.")
        }

        val active = records.filter { record ->
            period >= record.effectiveFrom && (record.effectiveTo == null || period <= record.effectiveTo)
        }
        if (active.isEmpty()) {
            return blocked("RGDU 2026 : régime de contribution logement et somme des taux éligibles à confirmer pour ${period.monthValue.toString().padStart(2, '0')}/${period.year}.")
        }
        if (active.size > 1) {
            return blocked("RGDU 2026 : plusieurs règles de coefficient se chevauchent ; calcul automatique bloqué.")
        }

        val selected = active.single()
        val standardDelta = when (selected.housingContributionRegime) {
            HousingContributionRegime.L813_5_1 -> T_DELTA_L813_5_1
            HousingContributionRegime.L813_5_2 -> T_DELTA_L813_5_2
        }
        val standardMaximum = T_MIN + standardDelta
        val effectiveMaximum = min(standardMaximum, selected.eligibleEmployerRateSum)
        val effectiveDelta = effectiveMaximum - T_MIN

        if (!effectiveDelta.isFinite() || effectiveDelta < 0.0) {
            return blocked("RGDU 2026 : somme des taux éligibles incompatible avec Tmin ; calcul automatique bloqué.")
        }

        return Snapshot(
            tDelta = effectiveDelta,
            maximumCoefficient = effectiveMaximum,
            source = selected.source.trim(),
            reliable = true,
            warnings = emptyList()
        )
    }

    fun isUsable(snapshot: Snapshot?): Boolean {
        if (snapshot == null || !snapshot.reliable || snapshot.source.isNullOrBlank()) return false
        val delta = snapshot.tDelta ?: return false
        val maximum = snapshot.maximumCoefficient ?: return false
        if (!delta.isFinite() || delta < 0.0 || !maximum.isFinite() || maximum < T_MIN || maximum > 1.0) return false
        return kotlin.math.abs((T_MIN + delta) - maximum) < 0.0000001
    }

    private fun validEligibleRateSum(value: Double): Boolean =
        value.isFinite() && value >= T_MIN && value <= 1.0

    private fun blocked(message: String) = Snapshot(
        tDelta = null,
        maximumCoefficient = null,
        source = null,
        reliable = false,
        warnings = listOf(message)
    )
}
